### PR TITLE
model: support gemma-3-it

### PR DESCRIPTION
**Pull request by mickqian**
## Motivation

Support gemma3-it.

FYI, `gemma3-1b-it` is an llm, `gemma3-pt` series are not chat models.
## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: sgl-kernel/csrc/moe/moe_align_kernel.cu [hunk 1]
```diff
@@ -108,23 +109,44 @@ void moe_align_block_size(
...
+  int threads;
+
+  if (num_experts <= 8) {
+    experts_per_warp = 8;
+    threads = 256;
+  } else if (num_experts <= 16) {
+    experts_per_warp = 16;
+    threads = 512;
+  } else {
+    experts_per_warp = WARP_SIZE;
```
Comment (yizhang2077) [inline sgl-kernel/csrc/moe/moe_align_kernel.cu]: remove changes about sgl_kernel

File: python/sglang/srt/entrypoints/http_server.py [hunk 2]
```diff
@@ -718,22 +718,22 @@ def _wait_and_warmup(
         ).tolist()
         json_data["sampling_params"]["max_new_tokens"] = 0
 
-    try:
-        for i in range(server_args.dp_size):
-            res = requests.post(
```
Comment (yizhang2077) [inline python/sglang/srt/entrypoints/http_server.py]: revert here

File: python/sglang/srt/managers/image_processors/base_image_processor.py [hunk 3]
```diff
@@ -118,87 +118,107 @@ def uniform_sample(l, n):
...
-        scaling_factor = min(1.0, MAX_NUM_FRAMES / total_frame_count)
+        scaling_factor = min(1.0, MAX_NUM_FRAMES / max(1, total_frame_count))
 
         assert len(image_data) == len(estimated_frames_list)
 
-        # Process each input with allocated frames
-        for image_index, (image, estimated_frames) in enumerate(
-            zip(image_data, estimated_frames_list)
-        ):
-            if len(all_frames) >= MAX_NUM_FRAMES:
```
Comment (yizhang2077) [inline python/sglang/srt/managers/image_processors/base_image_processor.py:156]: I think base image prcessor don't need change?

File: python/sglang/srt/managers/scheduler.py [hunk 4]
```diff
@@ -684,7 +684,7 @@ def handle_generate_request(
             if isinstance(req.finished_reason, FINISH_ABORT):
                 self._add_request_to_queue(req)
                 return
-
+        print(f"recv_req.image_inputs: {recv_req.image_inputs}")
```
Comment (yizhang2077) [inline python/sglang/srt/managers/scheduler.py]: remove print

File: python/sglang/srt/models/gemma2.py [hunk 5]
```diff
@@ -175,6 +178,14 @@ def forward(
         qkv, _ = self.qkv_proj(hidden_states)
         q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
         q, k = self.rotary_emb(positions, q, k)
+        global count
+        if first:
```
Comment (yizhang2077) [inline python/sglang/srt/models/gemma2.py]: remove count and print

File: python/sglang/srt/utils.py [hunk 6]
```diff
@@ -453,8 +453,10 @@ def load_image(image_file: Union[str, bytes]):
         image = Image.open(BytesIO(image_file))
     elif image_file.startswith("http://") or image_file.startswith("https://"):
         timeout = int(os.getenv("REQUEST_TIMEOUT", "3"))
-        response = requests.get(image_file, timeout=timeout)
-        image = Image.open(BytesIO(response.content))
+        image = Image.open(requests.get(image_file, stream=True).raw)
+
+        # response = requests.get(image_file, timeout=timeout)
```
Comment (yizhang2077) [inline python/sglang/srt/utils.py]: revert here?

File: python/sglang/srt/configs/qwen2_5_vl_config.py [hunk 7]
```diff
@@ -1002,5 +1002,5 @@ def make_batched_videos(videos) -> VideoInput:
         return BatchFeature(data=data, tensor_type=return_tensors)
 
 
-AutoImageProcessor.register(Qwen2_5_VLConfig, None, Qwen2_5_VLImageProcessor, None)
-AutoProcessor.register(Qwen2_5_VLConfig, Qwen2_5_VLProcessor)
+# AutoImageProcessor.register(Qwen2_5_VLConfig, None, Qwen2_5_VLImageProcessor, None)
```
Comment (yizhang2077) [inline python/sglang/srt/configs/qwen2_5_vl_config.py]: revert here

**Comment by mickqian** [comment]
An error occurred occasionally, in `test_vision_openai_server.TestGemma3itServer`:
```
[2025-03-14 13:07:28 TP0] Decode batch. #running-req: 1, #token: 596, token usage: 0.00, gen throughput (token/s): 36.60, largest-len: 0, #queue-req: 0, 
[2025-03-14 13:07:29 TP0] Decode batch. #running-req: 1, #token: 636, token usage: 0.00, gen throughput (token/s): 71.66, largest-len: 0, #queue-req: 0, 
[2025-03-14 13:07:29 TP0] Decode batch. #running-req: 1, #token: 676, token usage: 0.00, gen throughput (token/s): 71.60, largest-len: 0, #queue-req: 0, 
[2025-03-14 13:07:30 TP0] Decode batch. #running-req: 1, #token: 716, token usage: 0.00, gen throughput (token/s): 71.69, largest-len: 0, #queue-req: 0, 
[2025-03-14 13:07:30 TP0] Decode batch. #running-req: 1, #token: 756, token usage: 0.00, gen throughput (token/s): 70.71, largest-len: 0, #queue-req: 0, 
[2025-03-14 13:07:31 TP0] Decode batch. #running-req: 1, #token: 796, token usage: 0.00, gen throughput (token/s): 72.46, largest-len: 0, #queue-req: 0, 
[2025-03-14 13:07:31 TP0] Decode batch. #running-req: 1, #token: 836, token usage: 0.00, gen throughput (token/s): 73.02, largest-len: 0, #queue-req: 0, 
```

In this case, the generation never stops

**Comment by zhaochenyang20** [comment]
@mickqian @yizhang2077 is this ready to merge? approved?

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 8]
```diff
@@ -70,9 +71,6 @@ class ForwardMode(IntEnum):
     # It is now used for triggering the sampling_info_done event for the first prefill batch.
     DUMMY_FIRST = auto()
 
-    def is_prefill(self):
```
Comment (yizhang2077) [inline python/sglang/srt/model_executor/forward_batch_info.py]: Don't remove it

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 9]
```diff
@@ -331,6 +329,32 @@ def init_new(
 
         return ret
 
+    def get_merged_image_inputs(self) -> Optional[ImageInputs]:
```
Comment (yizhang2077) [inline python/sglang/srt/model_executor/forward_batch_info.py:332]: It seems unused?

File: python/sglang/srt/models/gemma3_causal.py [hunk 10]
```diff
@@ -0,0 +1,776 @@
...
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        gate_up, _ = self.gate_up_proj(x)
+        x = self.act_fn(gate_up)
+        x, _ = self.down_proj(x)
+        return x
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
+    """Applies Rotary Position Embedding to the query and key tensors.
+
```
Comment (yizhang2077) [inline python/sglang/srt/models/gemma3_causal.py]: Could you use `apply_rotary_pos_emb_vision` directly?

File: python/sglang/srt/models/gemma3_mm.py [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+
+cached_get_processor = lru_cache(get_processor)
+
+
+class Gemma3ImagePixelInputs(TypedDict):
+    pixel_values: torch.Tensor
+    """Shape: `(batch_size * num_images, num_channels, height, width)`"""
+
+
+class SiglipVisionLayer(nn.Module):
```
Comment (yizhang2077) [inline python/sglang/srt/models/gemma3_mm.py]: It seems unused?

File: python/sglang/srt/models/gemma3_causal.py [hunk 12]
```diff
@@ -0,0 +1,776 @@
...
+            )
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.model.embed_tokens
+
+    def set_input_embeddings(self, value):
+        self.model.embed_tokens = value
+
+    def get_output_embeddings(self):
```
Comment (yizhang2077) [inline python/sglang/srt/models/gemma3_causal.py]: some of get/set can be removed

File: python/sglang/srt/models/gemma3_mm.py [hunk 13]
```diff
@@ -0,0 +1,539 @@
...
+        kwargs["local_attn_masks"] = local_attn_masks
+        return kwargs
+
+    def get_input_embeddings(self):
+        return self.language_model.get_input_embeddings()
+
+    def set_input_embeddings(self, value):
+        self.language_model.set_input_embeddings(value)
+
+    def get_output_embeddings(self):
```
Comment (yizhang2077) [inline python/sglang/srt/models/gemma3_mm.py]: some of set/get can be removed

**Comment by mickqian** [comment]
It is abled to merged after #4451

**Comment by yizhang2077** [comment]
format code

**Comment by yizhang2077** [comment]
LGTM, once ci is passed, this pr can be merged. cc @zhaochenyang20

**Comment by Swipe4057** [comment]
Typically in PRs when adding models, accuracy is reported, for example, on MMLU implemented in the PR compared to the implementation in transformers, and for multimodal models, some multimodal benchmark. Could you advise how you verified the implementation?

**Comment by zhaochenyang20** [comment]
@yizhang2077 @mickqian 

This PR is depends on https://github.com/sgl-project/sglang/pull/4451

right?

**Comment by zhaochenyang20** [comment]
@mickqian fix the lint plz. I will merge this ASAP after transformers 4.48.3

**Comment by yizhang2077** [comment]
> Typically in PRs when adding models, accuracy is reported, for example, on MMLU implemented in the PR compared to the implementation in transformers, and for multimodal models, some multimodal benchmark. Could you advise how you verified the implementation?

@Swipe4057 we have a mmmu benchmark, and for each vlm model, we need verify this benchmark and compare with transformers implementation. Besides, if transformers supports this model, we suggest add an unittest to compare logits with hf.

**Comment by yizhang2077** [comment]
@mickqian plz paste mmmu benchmark result here.

**Comment by yizhang2077** [comment]
I add an issue for keeping track of current VLM models performance in mmmu benchmark. We can update benchmark result here https://github.com/sgl-project/sglang/issues/4456 @mickqian @zhaochenyang20

File: python/sglang/srt/managers/image_processors/base_image_processor.py [hunk 14]
```diff
@@ -122,76 +122,87 @@ def load_images(
...
             assert len(input_ids) and isinstance(input_ids[0], int)
             input_text = self._processor.tokenizer.decode(input_ids)
         else:
             input_text = input_ids
-
         if return_text:
-            text_parts = input_text.split(image_token)
+            import re
+
+            pattern = "(" + "|".join(re.escape(sep) for sep in [image_token]) + ")"
```
Comment (yizhang2077) [inline python/sglang/srt/managers/image_processors/base_image_processor.py:137]: Why we need add regex here?

**Review by yizhang2077 — APPROVED** [review]

**Comment by yizhang2077** [comment]
@zhaochenyang20 this PR can be merged

**Comment by mickqian** [comment]
@zhaochenyang20 This is ready. Many thanks

**Comment by zhaochenyang20** [comment]
@mickqian @yizhang2077 thanks. I will tell lianmin!

**Comment by zhaochenyang20** [comment]
@Ying1123 hey, ying, this can be merged. it's high-prioritized.

**Comment by Swipe4057** [comment]
@zhaochenyang20  @mickqian Hello, I've been running Gemma3 27b it model in vllm 0.8.0 and sglang installed from source with original weights on an H100 GPU. My results show that for the same long text-only query, the outputs of the models differ significantly. In vllm, generation proceeds normally, but in sglang, during long generation, it starts to degrade into garbage output and continues indefinitely. This phenomenon occurs with long queries and queries containing code. Could someone else test this as well?

Also, a question: is prefix caching supported in sglang for multimodal models, particularly Gemma 3?

**Comment by zhaochenyang20** [comment]
> @zhaochenyang20 @mickqian Hello, I've been running Gemma3 27b it model in vllm 0.8.0 and sglang installed from source with original weights on an H100 GPU. My results show that for the same long text-only query, the outputs of the models differ significantly. In vllm, generation proceeds normally, but in sglang, during long generation, it starts to degrade into garbage output and continues indefinitely. This phenomenon occurs with long queries and queries containing code. Could someone else test this as well?
> 
> Also, a question: is prefix caching supported in sglang for multimodal models, particularly Gemma 3?

supported. could you give us your reproducable scripts. We will fix this ASAP>

**Comment by Swipe4057** [comment]
@zhaochenyang20 
Here's my launch command:
image
Then ask a simple question:
How would you advise fixing high vulnerabilities in a Docker container that are found in the base Debian image?
Temperature 1 and top_p 0.95

You'll see that the generation doesn't stop and something like this will begin:
image

In the logs, the generation will continue:
image

**Comment by zhaochenyang20** [comment]
> @zhaochenyang20 Here's my launch command: image Then ask a simple question: How would you advise fixing high vulnerabilities in a Docker container that are found in the base Debian image? Temperature 1 and top_p 0.95
> 
> You'll see that the generation doesn't stop and something like this will begin: image
> 
> In the logs, the generation will continue: image

cc @mickqian mick

**Comment by mickqian** [comment]
a [fix](https://github.com/sgl-project/sglang/commit/5531f98f4db5cee32f9085c26bb82998ac5d1dc2) is on the way

**Comment by rangehow** [comment]
Gemma3's generation speed is surprisingly slow compared to other 3B/4B models like Qwen2.5-3B. Is the current Gemma3 implementation correct?

**Comment by zhaochenyang20** [comment]
great!
